app/controlplane/internal/biz/membership.go

```diff
@@ -41,21 +42,57 @@ type MembershipRepo interface {
 }
 
 type MembershipUseCase struct {
-	repo   MembershipRepo
-	logger *log.Helper
+	repo       MembershipRepo
+	orgUseCase *OrganizationUseCase
+	logger     *log.Helper
 }
 
-func NewMembershipUseCase(repo MembershipRepo, logger log.Logger) *MembershipUseCase {
-	return &MembershipUseCase{repo, log.NewHelper(logger)}
+func NewMembershipUseCase(repo MembershipRepo, orgUC *OrganizationUseCase, logger log.Logger) *MembershipUseCase {
+	return &MembershipUseCase{repo, orgUC, log.NewHelper(logger)}
 }
 
-func (uc *MembershipUseCase) Delete(ctx context.Context, id string) error {
-	uuid, err := uuid.Parse(id)
+// Delete deletes a membership from the database
+// and the associated org if the user is the only member
+func (uc *MembershipUseCase) DeleteWithOrg(ctx context.Context, userID, membershipID string) error {
+	membershipUUID, err := uuid.Parse(membershipID)
 	if err != nil {
 		return NewErrInvalidUUID(err)
 	}
 
-	return uc.repo.Delete(ctx, uuid)
+	userUUID, err := uuid.Parse(userID)
+	if err != nil {
+		return NewErrInvalidUUID(err)
+	}
+
+	// Check that the provided membershipID in fact belongs to a membership from the user
+	m, err := uc.repo.FindByIDInUser(ctx, userUUID, membershipUUID)
+	if err != nil {
+		return fmt.Errorf("failed to find membership: %w", err)
+	} else if m == nil {
+		return NewErrNotFound("membership")
+	}
+
+	uc.logger.Infow("msg", "Deleting membership", "user_id", userID, "membership_id", m.ID.String())
+	if err := uc.repo.Delete(ctx, membershipUUID); err != nil {
+		return fmt.Errorf("failed to delete membership: %w", err)
+	}
+
+	// Check number of members in the org
+	// If it's the only one, delete the org
+	membershipsInOrg, err := uc.repo.FindByOrg(ctx, m.OrganizationID)
+	if err != nil {
+		return fmt.Errorf("failed to find memberships in org: %w", err)
+	}
+
+	if len(membershipsInOrg) == 0 {
+		// Delete the org
```

Comment:
Just an architectural question, to make sure I understand everything well.

What happens if, in between the membership deletion and the org deletion, some other membership is created? Where are these races being handled?

Reply:
Good question; they are not. I thought about this one, and to solve it, we'll need to move most of this logic to the data layer, where we could leverage `ent transactions.` 

I just thought that the possibility of this race was very, very low and not worth the implementation in this first pass. That said, I'll create an issue for it so we don't forget.